File: src/main/java/com/cuupa/classificator/configuration/spring/ApplicationConfiguration.java
package com.cuupa.classificator.configuration.spring;

import com.cuupa.classificator.configuration.application.ApplicationProperties;
import com.cuupa.classificator.controller.ClassificatorController;
import com.cuupa.classificator.controller.GuiController;
import com.cuupa.classificator.services.Classificator;
import com.cuupa.classificator.services.kb.KnowledgeBaseExecutorService;
import com.cuupa.classificator.services.kb.KnowledgeBaseInitiator;
import com.cuupa.classificator.services.kb.KnowledgeManager;
import com.cuupa.classificator.services.stripper.PdfAnalyser;
import org.jetbrains.annotations.NotNull;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
public class ApplicationConfiguration {

    @NotNull
    @Bean
    public ClassificatorController classificatorController() {
        return new ClassificatorController(classificator());
    }

    @NotNull
    @Bean
    public GuiController guiController() {
        return new GuiController(classificator(), knowledgeManager());
    }

    @NotNull
    @Bean
    public Classificator classificator() {
        return new Classificator(knowledgeManager(), analyser());
    }

    @NotNull
    @Bean
    public KnowledgeManager knowledgeManager() {
        return new KnowledgeManager(knowledgeBaseInitiator(), knowledgeBaseExecutorService());
    }

    @NotNull
    @Bean
    public KnowledgeBaseExecutorService knowledgeBaseExecutorService() {
        return new KnowledgeBaseExecutorService();
    }

    @NotNull
    @Bean
    public KnowledgeBaseInitiator knowledgeBaseInitiator() {
        return new KnowledgeBaseInitiator(applicationProperties());
    }

    @NotNull
    @Bean
    public ApplicationProperties applicationProperties() {
        return new ApplicationProperties();
    }

    @NotNull
    @Bean
    public PdfAnalyser analyser() {
        return new PdfAnalyser();
    }
}

File: src/main/java/com/cuupa/classificator/configuration/spring/ApplicationConfiguration.kt
package com.cuupa.classificator.configuration.spring

import com.cuupa.classificator.configuration.application.ApplicationProperties
import com.cuupa.classificator.services.Classificator
import com.cuupa.classificator.services.kb.*
import com.cuupa.classificator.services.stripper.PdfAnalyser
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Configuration
open class ApplicationConfiguration {

    @Bean
    open fun classificator(): Classificator {
        return Classificator(knowledgeManager(), analyser())
    }

    @Bean
    open fun knowledgeManager(): KnowledgeManager {
        return KnowledgeManager(knowledgeBase(), knowledgeBaseExecutorService())
    }

    @Bean
    open fun knowledgeBaseExecutorService(): KnowledgeBaseExecutorService {
        return KnowledgeBaseExecutorService(topicService(), senderService(), metadataService())
    }

    @Bean
    open fun knowledgeBase(): KnowledgeBase {
        return knowledgeBaseInitiator().initKnowledgeBase()
    }

    @Bean
    open fun topicService(): TopicService {
        return TopicService(knowledgeBase().topicList)
    }

    @Bean
    open fun senderService(): SenderService {
        return SenderService(knowledgeBase().sendersList)
    }

    @Bean
    open fun metadataService(): MetadataService {
        return MetadataService(knowledgeBase().metadataList)
    }

    @Bean
    open fun knowledgeBaseInitiator(): KnowledgeBaseInitiator {
        return KnowledgeBaseInitiator(applicationProperties())
    }

    @Bean
    open fun applicationProperties(): ApplicationProperties {
        return ApplicationProperties()
    }

    @Bean
    open fun analyser(): PdfAnalyser {
        return PdfAnalyser()
    }
}